compiler/optimizer/optimizer.go

```diff
@@ -209,137 +321,123 @@ func (o *Optimizer) propagateScanOrder(op dag.Op, parent order.SortKey) (order.S
+	// use a semaphore here and let each possible path use the max concurrency.
+	if len(o.sources) == 0 {
+		return nil
 	}
-	if replicas > 50 {
-		// XXX arbitrary circuit breaker
-		return fmt.Errorf("parallelization factor too big: %d", n)
+	concurrency := n / len(o.sources)
+	if concurrency < 2 {
+		concurrency = 2
 	}
-	seq := o.entry
-	from, ok := seq.Ops[0].(*dag.From)
-	if !ok {
-		return nil
+	if concurrency > 50 {
+		// arbitrary circuit breaker
+		return fmt.Errorf("parallelization factor too big: %d", n)
 	}
-	trunks := poolTrunks(from)
-	if len(trunks) == 1 {
-		quietCuts(trunks[0])
-		if err := o.parallelizeTrunk(seq, trunks[0], replicas); err != nil {
+	for seq := range o.sources {
+		lister, slicer, rest := matchSource(seq.Ops)
+		// We parallelize the scanning to achieve the desired concurrency,
+		// then the step below pulls downstream operators into the parallel
+		// branches when possible, e.g., to parallelize aggregations etc.
+		parallel, err := o.parallelizeScan(rest, concurrency)
+		if err != nil {
 			return err
 		}
+		if parallel == nil {
+			// Leave the source path unmodified.
+			continue
+		}
+		front := []dag.Op{lister}
+		if slicer != nil {
+			front = append(front, slicer)
+		}
+		// Replace the source path with the parallelized gadget.
+		seq.Ops = append(front, parallel...)
 	}
+	o.findParPullups(o.entry)
+	removePassOps(o.entry)
 	return nil
 }
 
-func quietCuts(trunk *dag.Trunk) {
-	if trunk.Seq == nil {
-		return
-	}
-	for _, op := range trunk.Seq.Ops {
-		if cut, ok := op.(*dag.Cut); ok {
-			cut.Quiet = true
-		}
+func (o *Optimizer) lookupPool(id ksuid.KSUID) (*lake.Pool, error) {
+	if o.lake == nil {
+		return nil, errors.New("system error: lake operation cannot be used in non-lake context")
 	}
+	// This is fast because of the pool cache in the lake.
+	return o.lake.OpenPool(o.ctx, id)
 }
 
-func poolTrunks(from *dag.From) []*dag.Trunk {
-	var trunks []*dag.Trunk
-	for k := range from.Trunks {
-		trunk := &from.Trunks[k]
-		if _, ok := trunk.Source.(*dag.PoolScan); ok {
-			trunks = append(trunks, trunk)
-		}
+func matchSource(ops []dag.Op) (*dag.Lister, *dag.Slicer, []dag.Op) {
+	lister := ops[0].(*dag.Lister)
+	ops = ops[1:]
+	slicer, ok := ops[0].(*dag.Slicer)
+	if ok {
+		ops = ops[1:]
 	}
-	return trunks
-}
-
-func liftInto(trunk *dag.Trunk, branch []dag.Op) {
-	if trunk.Seq == nil {
-		trunk.Seq = &dag.Sequential{
-			Kind: "Sequential",
-		}
+	if _, ok := ops[0].(*dag.SeqScan); !ok {
+		panic("parseSource: no SeqScan")
 	}
-	trunk.Seq.Ops = append(trunk.Seq.Ops, branch...)
+	return lister, slicer, ops
 }
 
-func extend(trunk *dag.Trunk, op dag.Op) {
-	if trunk.Seq == nil {
-		trunk.Seq = &dag.Sequential{Kind: "Sequential"}
+// filterPushdown attempts to move any filter from the front of op chain
+// and returns the filter's expression (and the modified chain) so that
+// the runtime can push the filter predicate into the scanner.
+func filterPushdown(in []dag.Op) (dag.Expr, []dag.Op) {
```

Comment:
Nit: Give this a verby name. Maybe extractLeadingFilter?

Reply:
I'll call it `matchFilter`.  As we do more pattern matching in the optimizer, I think the `match*` pattern makes sense.